File: with-lombok/src/test/java/experimental/TestExtension.java
package experimental;

import static org.assertj.core.api.Assertions.*;

import org.junit.jupiter.api.Test;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.experimental.ExtensionMethod;

@ExtensionMethod({TestExtension.Extensions.class})
public class TestExtension {

    @Data
    @AllArgsConstructor
    public static class Person{
        String firstname;
        String lastname;
    }

    public static class Extensions {
        public static Person toPerson(String value){
            String[] split = value.split("\\|");
            return new Person(split[0], split[1]);
        }
    }

    @Test
    void testExtension() {
        Person person = "John|Doe".toPerson();

        assertThat(person.firstname).isEqualTo("John");
        assertThat(person.lastname).isEqualTo("Doe");
    }
}

File: with-kotlin/src/test/kotlin/TestExtension.kt
import kotlin.test.Test
import kotlin.test.assertEquals

class Person(val firstname:String, val lastname:String)

fun String.toPerson():Person{
    val (firstname, lastname) = this.split("|")
    return Person(firstname, lastname)
}

class TestExtension {
    @Test
    fun `should map a person`(){
        val person = "John|Doe".toPerson()

        assertEquals("John",person.firstname)
        assertEquals("Doe",person.lastname)
    }
}
